src/function/gds/frontier_morsel.cpp

```diff
@@ -0,0 +1,35 @@
+#include "function/gds/frontier_morsel.h"
+
+using namespace kuzu::common;
+
+namespace kuzu {
+namespace function {
+
+FrontierMorselDispatcher::FrontierMorselDispatcher(uint64_t maxThreads)
+    : maxOffset{INVALID_OFFSET}, maxThreads{maxThreads}, morselSize(UINT64_MAX) {
+    nextOffset.store(INVALID_OFFSET);
+}
+
+void FrontierMorselDispatcher::init(offset_t _maxOffset) {
+    maxOffset = _maxOffset;
+    nextOffset.store(0u);
+    // Frontier size calculation: The ideal scenario is to have k^2 many morsels where k
+    // the number of maximum threads that could be working on this frontier. However, if
+    // that is too small then we default to MIN_FRONTIER_MORSEL_SIZE.
+    auto idealMorselSize =
+        maxOffset / std::max(MIN_NUMBER_OF_FRONTIER_MORSELS, maxThreads * maxThreads);
+    morselSize = std::max(MIN_FRONTIER_MORSEL_SIZE, idealMorselSize);
+}
+
+bool FrontierMorselDispatcher::getNextRangeMorsel(FrontierMorsel& frontierMorsel) {
```

Comment:
Isn't the returned boolean flag kinda of carrying duplicated information with frontierMorsel? I think you should be able to find out if there is more morsel by checking if the returned morsel is valid or not.

Reply:
I'll refactor this with a different PR.